website/docs/r/cloudfront_distribution.html.markdown

```diff
@@ -142,6 +142,116 @@ resource "aws_cloudfront_distribution" "s3_distribution" {
 }
 ```
 
+### CloudFront distribution with custom origin
+
+The example below uses an S3 static web hosting origin for use-cases beyond asset downloads. AWS doesn't allow using the S3 website endpoint from CloudFront. This example works using a custom authentication and a custom origin to further provide caching and HTTPS.
+
+```hcl
+locals {
+  origin_id = "S3-Website-${aws_s3_bucket.website-bucket.bucket}"
+  cloudfront-authentication-user-agent = "somethingRandomEen8Ohta"
+}
+
+resource "aws_s3_bucket" "website-bucket" {
+  bucket = "webhosting-bucket-via-cloudfront"
+
+  website {
+    index_document = "index.html"
+  }
+}
+data "aws_iam_policy_document" "bucket-policy" {
+  statement {
+    actions   = ["s3:GetObject"]
+    resources = ["${aws_s3_bucket.website-bucket.arn}/*"]
+
+    principals {
+      type        = "*"
+      identifiers = ["*"]
+    }
+
+    condition {
+      test     = "StringEquals"
+      variable = "aws:UserAgent"
+      values   = ["${local.cloudfront-authentication-user-agent}"]
+    }
+  }
+}
+resource "aws_s3_bucket_policy" "bucket-policy" {
+  bucket = "${aws_s3_bucket.website-bucket.id}"
+
+  policy = "${data.aws_iam_policy_document.bucket-policy.json}"
+{
```

Comment:
I couldn't figure out how to make the new Github suggestion feature to also suggest removal of code, and it looks like you will need to remove the inline policy here.

Thank you for finding this method, by the way. Really saved me a lot of headache!

Reply:
Thank you, @austinbutler! I removed the residuals of the inline policy here-document.